Add deskintl (desktop international) layout set.

junaruga opened the pull request:
Let me send the pull request (PR) via GitHub for better visibility in this context. I can also send this patch by `git send-email` if you would like me to do it.

This PR adds a desktop layout set. I took over the work on the https://github.com/jjsullivan5196/wvkbd/pull/67.

The layout set is aimed at desktop, laptop, and tablet devices with a larger touchscreen. The layout is the US-International English below, adding a compose key.
https://en.wikipedia.org/wiki/QWERTY#US-International

I also referred to the [Framework Laptop 12](https://frame.work/laptop12)'s International English layout about the Fn (Function) layer's layout.

<img width="615" height="469" alt="framework-laptop-12-international-english" src="https://github.com/user-attachments/assets/6998e257-5d6e-46f3-a074-fa6f1410e793" />

The layout set was created based on the layout set by the user nine7nine (Jordan Johnston) <https://github.com/nine7nine>.

## Screenshots

The screenshots are below. I took the screenshots on the [Sway window manager](https://swaywm.org/) on Fedora Linux 42 on Framework Laptop 12 (a 2-in-1 laptop with a touchscreen). You can download the original image files [here](https://drive.google.com/drive/folders/1xuy4-H4fsJaevShItrUpW_lI3oFK0mvT?usp=drive_link) on my Google Drive.

### Landscape

**Layout `Landscape` (`keys_landscape`)**
<img width="640" height="400" alt="wvkbd-desktop-landscape" src="https://github.com/user-attachments/assets/8d943f64-9e7d-4ba4-9458-c880384483f5" />

**Layout `LandscapeSpecial` (`keys_landscape_special`)**
<img width="640" height="400" alt="wvkbd-desktop-landscape-special" src="https://github.com/user-attachments/assets/721fc105-5463-48f0-a982-598e82e245fa" />

### Portrait

**Layout `Full` (`keys_full`)**
<img width="400" height="640" alt="wvkbd-desktop-portrait" src="https://github.com/user-attachments/assets/acc5792e-b06e-4000-8f96-3087bc09f828" />

**Layout `Special` (`keys_special`)**
<img width="400" height="640" alt="wvkbd-desktop-portrait-special" src="https://github.com/user-attachments/assets/b0408bf4-1880-4223-852b-c9f68b5e07ea" />

## Reply to the comments

@proycon's comment (https://github.com/jjsullivan5196/wvkbd/pull/67#issuecomment-2027782803):

> >  Just credit my part of it in the commit, and we are good!
> This was my first thought as well; I understand this
> comes from [@nine7nine](https://github.com/nine7nine)'s fork with additional improvements by [@FearlessSpiff](https://github.com/FearlessSpiff) on
top. Right now the commit history in this PR doesn't reflect this though.
...
> [@FearlessSpiff](https://github.com/FearlessSpiff): Could you rebase the PR so the original commits are as
> they were (carrying the proper Author for attribution)? And then yours
> on top of that? (you might still have them in your working tree)

I added @nine7nine's name as a credit in the commit message. I couldn't find @nine7nine's branch in the forked repository. So, this PR is based on https://github.com/jjsullivan5196/wvkbd/pull/67, and rebased on the latest master branch. And I modified and adjusted the code. 

> Some other comments:
>
> I didn't really notice the line is not straight, but I do notice the
> slight border on the left/right yes. I suppose it's a purely aestethic
> choice or does it have some functional reason too?

I adjusted each key's width with `0.25 * N`. I set each row's total width as 14.5 in each layout.

> * I see some of the keys were renamed, which is okay but care has to be
>   taken to do  it consistently for all layers in the desktop layout
>   then. Changing labels also makes it a bit harder to document keys clearly as
>   the README refers to Cmp for example, I'd suggest not renaming these
>   keys unless there's a good reason). I noticed:
>     * "Cmp" changed to "><"
>     * Sup changed to "<Arch logo>"; ideally I think we better keep a distro-neutral label
>       in wvkbd rather than an Arch/Debian/Gentoo/.. logo
>       (even though I too mostly use Arch, btw)

I fixed the above things.

> * though 99% percent of desktop users are probably on a landscape
>   layout, some 1% have a portrait monitor. The layout that opens
>   then is still the same as in wvkbd-mobintl. Is this by choice? I'd be inclined to
>   just use the same new landscape layout in portrait mode since the screen is probably
>   wide enough on desktop.

I fixed the above thing. Now, the landscape and portrait modes have the same layouts.

> * Perhaps an AltGr key would be useful on the default layer (right of spacebar)?
>   It would give alternative access to some special keys users may
>   already be used too (and as an alterntive for Cmp).

I fixed the above thing. I added the AltGr key.

> * Currently the keymaps and layouts for the other languages are still in
>   the code, but not really modified. If they're not used they should be
>   removed. Ideally though, a desktop layout for cyrillic etc would be
>   great (if anyone's up for it).

I removed unused keymaps and layouts for other languages, also some compose layouts (`ComposeCyr*`) to simplify the layout set. I wanted to show a simple example so that users can modify and extend the `layout.*.h` file easily.

I see some duplicated codes between `layout.mobintl.h` and `layout.deskintl.h`. Can we move the code related to the layouts `Compose*` into a header file, such as `layout_compose.h`, and the `layout.mobintl.h` and `layout.deskintl.h` load the `layout_compose.h`?

@justinlovinger's comment (https://github.com/jjsullivan5196/wvkbd/pull/67#issuecomment-2645300798):

> `<`, `>`, and `?` type the wrong symbols.

I still have no idea about how to fix this issue. Could you tell me how to fix it?

## Notes

I executed `make format`. However, after executing the command, I see many changes including the `keymap.mobintl.h` and `layout.mobintl.h`. So, I didn't apply the changes to `keymap.deskintl.h` and `layout.deskintl.h`. Because I think it's better to align the content between the `*.mobintl.h` and `*.deskintl.h`.

```
$ clang-format --version
clang-format version 20.1.8 (Fedora 20.1.8-3.fc42)

$ make format

$ git status
On branch wip/layout-desktop
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   config.def.h
	modified:   drw.c
	modified:   drw.h
	modified:   keyboard.c
	modified:   keyboard.h
	typechange: keymap.deskintl.h
	modified:   keymap.mobintl.h
	modified:   layout.deskintl.h
	modified:   layout.mobintl.h
	modified:   main.c
	modified:   os-compatibility.h
```

So, what do you think?

proycon commented:
Thanks, great work! This looks nice and a good addition to have for all
those wanting a better desktop layout!

On Fri Aug 8, 2025 at 9:28 AM CEST, Jun Aruga wrote:
> I added @nine7nine's name as a credit in the commit message. I
> couldn't find @nine7nine's branch in the forked repository.
> So, this PR is based on
> https://github.com/jjsullivan5196/wvkbd/pull/67, and rebased on the
> latest master branch. And I modified and adjusted the code. 

Right, perhaps they removed their original branch, the message in the
commit message looks fine.

> I adjusted each key's width with `0.25 * N`. I set each row's total width as 14.5 in each layout.

The sizing looks good! Keys seem natural and sensible as they're close to most hardware keyboard layouts

> > * I see some of the keys were renamed, which is okay but care has to be
> >   taken to do  it consistently for all layers in the desktop layout
> >   then. Changing labels also makes it a bit harder to document keys clearly as
> >   the README refers to Cmp for example, I'd suggest not renaming these
> >   keys unless there's a good reason). I noticed:
> >     * &quot;Cmp&quot; changed to &quot;&gt;&lt;&quot;
> >     * Sup changed to &quot;&lt;Arch logo&gt;&quot;; ideally I think we better keep a distro-neutral label
> >       in wvkbd rather than an Arch/Debian/Gentoo/.. logo
> >       (even though I too mostly use Arch, btw)
>
> I fixed the above things.

Perfect!

>
>  Now, the landscape and portrait modes have the same layouts.

Yep, I can confirm it works on my desktop portrait monitor too. I do
have a bug on my landscape monitor for some reason (keyboard appears
but hides immediately), but that can't really be due to this patch but
must be caused elsewhere, I'll try to pinpoint it.

> I added the AltGr key.

Nice!

> I removed unused keymaps and layouts for other languages, also some
> compose layouts (`ComposeMath`, `ComposePunctuation`,
> `ComposeBracket`, and `ComposeCyr*`) to simplify the layout set. I
> wanted to show a simple example so that users can modify and extend
> the `layout.*.h` file easily.

Yes, fair enough. This is a good basis upon which to extend further.
Would be great to have this for the other languages as well eventually,
especailly because the name still suggest 'international'. 

I noticed you now use Fn to move to the next layer, which makes sense
with the current layouts. Technically on hardware keyboards Fn is a
modifier key though. So I'm not sure what's best yet, Fn or the just the
keyboard icon as we use on the mobile one (for consistency). I'm also
thinking ahead to situations when we do have layouts like cyrillic and
arabic again. What do you think?

> I see some duplicated codes between `layout.mobintl.h` and
> `layout.deskintl.h`. Can we move the code related to the layouts
> `Compose*` into a header file, such as `layout_compose.h`, and the
> `layout.mobintl.h` and `layout.deskintl.h` load the
> `layout_compose.h`?

Hmm.. perhaps yes.. On the other hand though, the idea was that these
compose layers might alsobe  different per layout (to keep consistency
with the primary layout), even though you haven't made use of that that
here. The keyboard now switches between 6 and 4 rows when opening a
compose layer, making the compose keys bigger, all that's not a very big
deal though as it works fine still.

> @justinlovinger's comment (https://github.com/jjsullivan5196/wvkbd/pull/67#issuecomment-2645300798):
>
> &gt; `&lt;`, `&gt;`, and `?` type the wrong symbols.
>
> I still have no idea about how to fix this issue. Could you tell me how to fix it?

It's in the keymap, the keymap was modified in a few places to fit the
mobile keyboard. So we'll have to copy rather than symlink the
keymap.*.h file and make some minor changes. I'll take a look.

> So, what do you think?

Great work! I'll have a look at some of the things I mentioned later
and then merge.

proycon commented:
Short follow up on the unrelated bug I found:

> I do have a bug on my landscape monitor for some reason (keyboard appears
> but hides immediately), but that can't really be due to this patch but
> must be caused elsewhere, I'll try to pinpoint it.

So this was unrelated to this patch and specific to a multi-monitor setup, I pinpointed it and proposed a patch for it here: https://lists.sr.ht/~mil/sxmo-devel/patches/61264

junaruga commented:
First, I rebased this PR, updating the commit message too. Below are the updated screenshots. I also updated the screenshots on [my Google Drive](https://drive.google.com/drive/folders/1xuy4-H4fsJaevShItrUpW_lI3oFK0mvT?usp=drive_link).

Landscape
<img width="640" height="400" alt="wvkbd-desktop-landscape" src="https://github.com/user-attachments/assets/015a3817-582d-464b-a85c-5c74b02f4df7" />

<img width="640" height="400" alt="wvkbd-desktop-landscape-special" src="https://github.com/user-attachments/assets/42696c6d-ba37-4ea4-bc3f-d9c7048fad23" />

Portrait
<img width="400" height="640" alt="wvkbd-desktop-portrait" src="https://github.com/user-attachments/assets/2dfb5df6-ec20-4525-bbec-0f9f06c045b8" />

<img width="400" height="640" alt="wvkbd-desktop-portrait-special" src="https://github.com/user-attachments/assets/2cd9d9ea-1cb4-49d9-8a16-27d5d674b757" />

Let me comment inline.

> Ok, we can always re-add that it in a later stage. I suppose especially the emoji
keyboard may be interesting for users eventually.

All right! I thought the deskintel layout set was to emulate what we can do with a physical keyboard. And there is no feature to type emoji codes directly from the physical keyboard. So, I didn't think the emoji keyboard was essential for the deskintel layout set. But if people are interested in the emoji keyboard, we can add it later.

> Yes, no rush there either, if the foundation is solid we can build
upon it later.

Okay!

> Yes, I think adding a key with keyboard icon to change layout is best, possibly in
addition to the Fn button doing what you suggest (use key_type Layout
for that). Regarding 'somewhere', I would be picky there: I think the
keyboard button should be the always be the bottom-left key (or the
next one) on all layouts, because that enables users to quickly cycle through all
layouts. The challenge is getting the alignment to look nice in relation
to the other keys if you insert an extra key there.

Sure. I added the keyboard button with the keyboard icon at the position of the bottom-left in the 4 layouts: Full, Special, Landscape, and LandscapeSpecial, to show what the new layouts look like. As we don't have other language layouts yet, the keyboard button moves to the Index layout. We can change the behavior to move to another language layout when we add other language layouts.

What do you think about the new layouts with the keyboard button?

```
{"⌨͕", "⌨͔", 0.75, Layout, 0, &layouts[Index], .scheme = 1},
```

> On the landscapeSpecial layer you can also consider renaming Fn to Abc
(since that's essentialy the toggle back then). We do this on the mobile
version of the layout too.

Okay. I renamed the Fn to Abc with `BackLayer` like this.

```
{"Abc", "Abc", 0.75, BackLayer, .scheme = 1},
```

> Even though you changed opinion, small note here anyway: the actual
order of layers can always be overriden by the user at runtime (-l and
-L parameters).

All right. I changed the Fn keys' setting on Landscape and Full layout from `NextLayer` to `&layouts[LandscapeSpecial]` or `&layouts[Special]` to specify the destination layout in the case of the actual order of the layers being overridden by the command parameters.

Landscape layout:

```
{"Fn", "Fn", 0.75, Layout, 0, &layouts[LandscapeSpecial], .scheme = 1},
```

Full layout:

```
{"Fn", "Fn", 0.75, Layout, 0, &layouts[Special], .scheme = 1},
```

> Yes, I can imagine it's a bit overwhelming at first. I agree there's
potential for splitting out things here, but perhaps best leave that for a separate PR?

Yes, splitting the layout header files can be a separate PR. I don't intend to split the layout header files on this pull PR. :)

> You'll want to make a copy instead of symlink of `keymap.mobintl.h` and then on line 1229 do:

Thank you for investigating the issue. I applied the reverse of the `keymap.h` in the patch. Now the difference between the `keymap.mobintl.h` and `keymap.deskintl.h` is below.

```
$ diff -u keymap.mobintl.h keymap.deskintl.h
--- keymap.mobintl.h	2025-08-08 20:47:26.449518525 +0200
+++ keymap.deskintl.h	2025-08-09 07:34:14.443834192 +0200
@@ -1226,9 +1226,9 @@
         key <AB05>               {	[               b,               B, equal ] };\
         key <AB06>               {	[               n,               N, question ] };\
         key <AB07>               {	[               m,               M, exclam ] };\
-        key <AB08>               {	[           comma,      apostrophe, less, U00AB] };\
-        key <AB09>               {	[          period,        question, greater, U00BB] };\
-        key <AB10>               {	[           slash,        greater ] };\
+        key <AB08>               {	[           comma,            less, backslash] };\
+        key <AB09>               {	[          period,         greater, bar ] };\
+        key <AB10>               {	[           slash,        question ] };\
         key <I147>               {  [      exclamdown,   questiondown, exclamdown ] };\
         key <RTSH>               {	[         Shift_R ] };\
         key <KPMU>               {\
```


> Here too we might perhaps later reorganize things a bit to have a bit
less duplication. They keymaps are even more overwhelming than the
layouts :)

That's amazing if we reorganize the things later! Thank you!

In addition, I also found the following keys type the wrong characters. Maybe we need to change the `keymap.deskintl.h`. Is there an easy way to test the key typing, such as `sudo showkey`, in the case of testing a physical keyboard? I see that the `sudo showkey` is useful to confirm the key codes when using a physical keyboard. But in the case of the virtual keyboard, how to confirm a typed key's key code? I haven't tested the cases of the Shift + F1..F12 keys, but I suspect the keys may type the wrong characters.

If you know how to fix the `keymap.deskintl.h`, could you tell me how to fix it?

* Shift+Space
* Shift+Up arrow
* Shift+Down arrow
* Shift+Left arrow
* Shift+Right arrow
* Shift+Del
* Shift+F1~F12 How to test?

junaruga commented:
> Sure. I added the keyboard button with the keyboard icon at the position of the bottom-left in the 4 layouts: Full, Special, Landscape, and LandscapeSpecial, to show what the new layouts look like.

The current design is below. The challenge is the size of the keyboard, and and Fn keys are a bit small with a width of 0.75, while the left shift key is accessible.

* 5th row left: Left shift key (width 2.5), 
* 6th row left: Keyboard key (width 0.75), Fn key (width 0.75), Left Ctr key (width 1.0)

I think the following designs are also good. What do you think?

* 5th row: Fn, key (width 1.0), Left shift key (width 1.5), 
* 6th row: Keyboard key (width 1.5), Left Ctr key (width 1.0)

or

* 5th row: Fn, key (width 1.0), Left shift key (width 1.5), 
* 6th row: Keyboard key (width 1.0), Left Ctr key (width 1.5)


What do you think?

junaruga commented:
> In addition, I also found the following keys type the wrong characters. Maybe we need to change the `keymap.deskintl.h`. Is there an easy way to test the key typing, such as `sudo showkey`, in the case of testing a physical keyboard? I see that the `sudo showkey` is useful to confirm the key codes when using a physical keyboard. But in the case of the virtual keyboard, how to confirm a typed key's key code? I haven't tested the cases of the Shift + F1..F12 keys, but I suspect the keys may type the wrong characters.
> 
> If you know how to fix the `keymap.deskintl.h`, could you tell me how to fix it?
> 
> * Shift+Space
> * Shift+Up arrow
> * Shift+Down arrow
> * Shift+Left arrow
> * Shift+Right arrow
> * Shift+Del
> * Shift+F1~F12 How to test?

I am also not sure if the following key's behaviors are correct.

* Shift + Del
* Shift + Ins
* Shift + Home
* Shift + End
* Shift + PgUp
* Shift + PgDn

I am seeing the following settings.

keymap.deskintl.h

```
xkb_symbols \"(unnamed)\" {\
        name[group1]=\"wvkbd\";\
\
...
        key <SPCE>               {      [           space ] };\
...
        key <FK01>               {\
                type= \"CTRL+ALT\",\
                symbols[Group1]= [              F1,              F1,              F1,              F1, XF86Switch_VT_1 ]\
        };\
...
        key <FK10>               {\
                type= \"CTRL+ALT\",\
                symbols[Group1]= [             F10,             F10,             F10,             F10, XF86Switch_VT_10 ]\
        };\
...
        key <HOME>               {      [            Home ] };\
        key <UP>                 {      [              Up ] };\
        key <PGUP>               {      [           Prior ] };\
        key <LEFT>               {      [            Left ] };\
        key <RGHT>               {      [           Right ] };\
        key <END>                {      [             End ] };\
        key <DOWN>               {      [            Down ] };\
        key <PGDN>               {      [            Next ] };\
        key <INS>                {      [          Insert ] };\
        key <DELE>               {      [          Delete ] };\
...
};\

```

Is it better to set the second level (the key code with shift) and third level (the key code with alt) explicitly like this, which is like the "a" key's setting?

```
        key <SPCE>               {      [           space, space, space ] };\
```

```
        key <AC01>               {      [               a,               A, minus ] };\
```

proycon commented:
On Sat Aug 9, 2025 at 1:33 PM CEST, Jun Aruga wrote:
> I am also not sure if the following key's behaviors are correct.
>
> * Shift + Del
> * Shift + Ins
> * Shift + Home
> * Shift + End
> * Shift + PgUp
> * Shift + PgDn
>
> I am seeing the following settings.
>
> keymap.deskintl.h
>
> ```
> xkb_symbols \"(unnamed)\" {\
>         name[group1]=\"wvkbd\";\
> \
> ...
>         key <SPCE>               {      [           space ] };\
> ...
>         key <FK01>               {\
>                 type= \"CTRL+ALT\",\
>                 symbols[Group1]= [              F1,              F1,              F1,              F1, XF86Switch_VT_1 ]\
>         };\
> ...
>         key <FK10>               {\
>                 type= \"CTRL+ALT\",\
>                 symbols[Group1]= [             F10,             F10,             F10,             F10, XF86Switch_VT_10 ]\
>         };\
> ...
>         key <HOME>               {      [            Home ] };\
>         key <UP>                 {      [              Up ] };\
>         key <PGUP>               {      [           Prior ] };\
>         key <LEFT>               {      [            Left ] };\
>         key <RGHT>               {      [           Right ] };\
>         key <END>                {      [             End ] };\
>         key <DOWN>               {      [            Down ] };\
>         key <PGDN>               {      [            Next ] };\
>         key <INS>                {      [          Insert ] };\
>         key <DELE>               {      [          Delete ] };\
> ...
> };\
>
> ```
>
> Is it better to set the second level (the key code with shift) and third level (the key code with alt) explicitly like this, which is like the "a" key's setting?
>
> ```
>         key <SPCE>               {      [           space, space, space ] };\
> ```
>
> ```
>         key <AC01>               {      [               a,               A, minus ] };\
> ```

They keymap is pretty standard on this point, unless we want really
specific behaviour (like pageup on Shift+Up) I don't think it needs to
go in there.

But you are indeed on to something that the behaviour seems different
with wvkbd than with a hardware keyboard, and I wonder if it's only
due to Shift_L vs Shift_R or if there's more to it. Perhaps we should
leave this too for a later issue/PR as it's something that affects
both layouts.

proycon commented:
On Sat Aug 9, 2025 at 1:57 PM CEST, Jun Aruga wrote:
> One more question. Should we remove the setting for the other
> languages in the "cyrillic", "arabic", "georgian", "greek",
> "persian", "hebrew" in the `keymap.deskintl.h` in this PR, right?

Yes, we can re-add them later once layouts are there.

junaruga commented:
First, I rebased this PR, updating the commit message. Below are the updated screenshots. I also updated the screenshots on [my Google Drive](https://drive.google.com/drive/folders/1xuy4-H4fsJaevShItrUpW_lI3oFK0mvT?usp=drive_link).

Landscape
<img width="640" height="400" alt="wvkbd-desktop-landscape" src="https://github.com/user-attachments/assets/d61ab7c9-8024-45ea-be9d-2b76de6b5993" />
<img width="640" height="400" alt="wvkbd-desktop-landscape-special" src="https://github.com/user-attachments/assets/a2a7feee-8000-45a3-88df-5dc249ab45e6" />

Portrait
<img width="400" height="640" alt="wvkbd-desktop-portrait" src="https://github.com/user-attachments/assets/fcd0f8d7-d1e7-46a4-82b6-2eaa22d1660a" />
<img width="400" height="640" alt="wvkbd-desktop-portrait-special" src="https://github.com/user-attachments/assets/03d4ca4b-3174-467e-bc68-3d447ff974cd" />

Let me comment inline below.

> I'd rather tie it to NextLayer already, even though that's technically the same as Fn now. Just for consistency future expansion. The index is always accessible with Cmp + keyboard button.
> ...
> Reconsidering everything, and with my previous comment in mind, I'd say the Fn key serves no purpose anymore on the primary layout and can just be removed? It would be simplest and cleanest.

All right. Now the keyboard icon key is below. I removed the Fn key.

```
{"⌨͕", "⌨͔", 1.5, NextLayer, .scheme = 1},
```

> Now you mention this, there's another thing I missed in the first review:
> the landscape and portrait layouts are the exactly the same! That is
perfectly fine for a desktop keyboard since there's enough space available.
> They were separated for the mobile variant because we had to really make
things fit with the tighter constraints.
> ..snip..
> That removes a fair bit of duplication and makes things easier.

Oh, that's a great suggestion. I added the commit to support a layout-specific config.h at the first commit. Note that I also added `config.h` to the clean task in the `Makefile` at the first commit. Because when we don't remove the `config.h` in the clean task, the following error can happen.

```
$ ls config.h
ls: cannot access 'config.h': No such file or directory

$ make

$ ls -l config.h
-rw-r--r--. 1 jaruga jaruga 1286 Aug  9 19:40 config.h

$ make clean

$ ls -l config.h
-rw-r--r--. 1 jaruga jaruga 1286 Aug  9 19:40 config.h

$ make clean LAYOUT=deskintl

$ ls -l config.h # This file is still for mobintl
-rw-r--r--. 1 jaruga jaruga 1286 Aug  9 19:40 config.h

$ make LAYOUT=deskintl
...
In file included from main.c:20:
config.h:40:3: error: ‘Landscape’ undeclared here (not in a function)
   40 |   Landscape, // First layout is the default layout on startup
      |   ^~~~~~~~~
config.h:41:3: error: ‘LandscapeSpecial’ undeclared here (not in a function)
   41 |   LandscapeSpecial,
      |   ^~~~~~~~~~~~~~~~
make: *** [<builtin>: main.o] Error 1
```

The second commit is to add the deskintl layout set. I removed the duplicated codes in the `layout.deskintl.h`.

> It currently produces a TAB and is done by separate code introduced in
https://github.com/jjsullivan5196/wvkbd/commit/29e32edf5c64984c4adeb9c3ec4a68fb833465b3 . I suppose we can let the
preprocesser handle it to turn this behaviour on/off at compile down via
directive in config.h. I can solve it in a separate commit afterwards if
you prefer.

Okay. Handling the behavior on/off in the config.h looks good idea. It can be a separate commit. Thank you for that.

> > * Shift+Up arrow
> > * Shift+Down arrow
> > * Shift+Left arrow
> > * Shift+Right arrow
> > * Shift+Del
> I don't think these were really changed. What behaviour do you expect
there? When I check with `wev` I see the difference only that my hardware
keyboard sends Shift_L or Shift_R and wvkbd just does Shift.

For example, when I typed Shift + Up key on the `wvkbd-deskintl` on the terminal [alacritty](https://alacritty.org/), "A" was typed. However, I noticed this didn't happen on Firefox. So, this may be the alacritty-specific issue. We can ignore this issue for now.

> With wev?

Thanks for the info!

> They keymap is pretty standard on this point, unless we want really
specific behaviour (like pageup on Shift+Up) I don't think it needs to
go in there.
>
> But you are indeed on to something that the behaviour seems different
with wvkbd than with a hardware keyboard, and I wonder if it's only
due to Shift_L vs Shift_R or if there's more to it. Perhaps we should
leave this too for a later issue/PR as it's something that affects
both layouts.

Yes, I agree with it. I am actually happy with the current keymap.

> Yes, we can re-add them later once layouts are there.

I removed the setting for the other languages in the "cyrillic", "arabic", "georgian", "greek", "persian", "hebrew" in the `keymap.deskintl.h`.

I think I fixed all the things mentioned in the review. What do you think?

proycon commented:
Thanks! All looks good now! Merged!

> I added the commit to support a layout-specific config.h at the first commit. Note that I also added config.h to the clean task in the Makefile at the first commit. Because when we don't remove the config.h in the clean task, the following error can happen.

Good point, a `make clean` is indeed needed between compiling different layouts.

> For example, when I typed Shift + Up key on the wvkbd-deskintl on the terminal [alacritty](https://alacritty.org/), "A" was typed. However, I noticed this didn't happen on Firefox. So, this may be the alacritty-specific issue. We can ignore this issue for now.

Doesn't happen on foot either. Perhaps the A comes from `^[[A` which is what is send on up arrow.

junaruga commented:
> Thanks! All looks good now! Merged!

Thank you for your review!

> > For example, when I typed Shift + Up key on the wvkbd-deskintl on the terminal [alacritty](https://alacritty.org/), "A" was typed. However, I noticed this didn't happen on Firefox. So, this may be the alacritty-specific issue. We can ignore this issue for now.
> 
> Doesn't happen on foot either. Perhaps the A comes from `^[[A` which is what is send on up arrow.

Okay. When I typed Shift + Up on the physical keyboard, the alacritty typed "A". So, this is not a wvkbd issue.

By the way, related to the first commit on this PR (49975e78ee437c6fb4a5f6ea3ccf8bc90dc65536), I noticed that I missed updating the `README.md`. The `config.def.h` is still written in the `README.md`. Maybe I will send a patch later.

junaruga commented:
> By the way, related to the first commit on this PR ([49975e7](https://github.com/jjsullivan5196/wvkbd/commit/49975e78ee437c6fb4a5f6ea3ccf8bc90dc65536)), I noticed that I missed updating the `README.md`. The `config.def.h` is still written in the `README.md`. Maybe I will send a patch later.

I sent the PR https://github.com/jjsullivan5196/wvkbd/pull/104 for the above thing.
